core/src/main/java/io/grpc/Deadline.java

```diff
@@ -0,0 +1,136 @@
+
+  private static final AtomicIntegerFieldUpdater<Deadline> expiredUpdater =
+      AtomicIntegerFieldUpdater.newUpdater(Deadline.class, "expired");
+
+  /**
+   * Create a deadline that will expire at the specified offset from the current system clock.
+   * @param duration A non-negative duration.
+   * @param units The time unit for the duration.
+   * @return A new deadline.
+   */
+  public static Deadline after(long duration, TimeUnit units) {
+    Preconditions.checkNotNull(units);
+    if (duration > 0) {
+      return new Deadline(System.nanoTime(), units.toNanos(duration), false);
+    } else {
+      // Deadlines at the current instant or earlier are already expired.
+      return new Deadline(System.nanoTime(), units.toNanos(duration), true);
+    }
+  }
+
+  private final long deadlineNanos;
+
+  @SuppressWarnings("unused")
+  private volatile int expired;
+
+  private Deadline(long baseInstant, long offset, boolean baseInstantAlreadyExpired) {
+    if (offset > 0 && Long.MAX_VALUE - offset < baseInstant) {
+      deadlineNanos = Long.MAX_VALUE;
+      expired = 0;
+    } else if (offset < 0 && Long.MIN_VALUE - offset > baseInstant) {
+      deadlineNanos = Long.MIN_VALUE;
+      expired = 1;
+    } else {
+      deadlineNanos = baseInstant + offset;
+      expired = baseInstantAlreadyExpired && offset <= 0 ? 1 : 0;
+    }
+  }
+
+  /**
+   * Has this deadline expired
+   * @return {@code true} if it has, otherwise {@code false}.
+   */
+  public boolean isExpired() {
+    if (expiredUpdater.get(this) == 0) {
+      if (deadlineNanos <= System.nanoTime()) {
+        expiredUpdater.set(this, 1);
+      } else {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Is {@code this} deadline before another.
+   */
+  public boolean isBefore(Deadline other) {
+    return this.deadlineNanos < other.deadlineNanos;
+  }
+
+  /**
+   * Return the minimum deadline of {@code this} or an other deadline.
+   * @param other deadline to compare with {@code this}.
+   */
+  public Deadline minimum(Deadline other) {
+    return isBefore(other) ? this : other;
+  }
+
+  /**
+   * Create a new deadline that is offset from {@code this}.
+   */
+  public Deadline offset(long offset, TimeUnit units) {
+    // May already be expired
+    if (offset == 0) {
+      return this;
+    }
+    return new Deadline(deadlineNanos,  units.toNanos(offset), expiredUpdater.get(this) == 1);
+  }
+
+  /**
+   * Schedule a task to be run when the deadline expires.
+   * @param task to run on expiration
+   * @param scheduler used to execute the task
+   * @return {@link ScheduledFuture} which can be used to cancel execution of the task
+   */
+  public ScheduledFuture<?> runOnExpiration(Runnable task, ScheduledExecutorService scheduler) {
+    Preconditions.checkNotNull(task, "task");
+    Preconditions.checkNotNull(scheduler, "scheduler");
+    return scheduler.schedule(task, Math.max(0, deadlineNanos - System.nanoTime()),
```

Comment:
Note that the max() is unnecessary since negative delays are fine:
> Zero and negative delays (but not periods) are also allowed in schedule methods, and are treated as requests for immediate execution.